CRUD операции с лайками.

no-vi-ka opened the pull request:
Реализованы операции создания, изменения, удаления лайков пользователем. Получение осуществляется по id пользователя - просмотр всех его лайков.
Не трогала events в плане подсчёта рейтинга события и добавления поля с рейтингом в модель.

no-vi-ka committed:
CRUD операции с лайками.

no-vi-ka committed:
checkstyle

no-vi-ka committed:
checkstyle

elbonreedZ commented on main-service/src/main/java/ru/practicum/rating/dto/NewRatingDto.java:
```diff
@@ -0,0 +1,27 @@
+public class NewRatingDto {
+    @NotNull(message = "Поле user должно быть указано.")
+    @NotBlank(message = "Поле user не должно быть пустым.")
+    User user;
```
это поле здень не нужно, так так userId в pathvariable передается

elbonreedZ commented on main-service/src/main/java/ru/practicum/rating/dto/NewRatingDto.java:
```diff
@@ -0,0 +1,27 @@
+    User user;
+    @NotNull(message = "Поле event должно быть указано.")
+    @NotBlank(message = "Поле event не должно быть пустым.")
+    Event event;
```
ну а тут только event_id

elbonreedZ commented on main-service/src/main/java/ru/practicum/rating/dto/NewRatingDto.java:
```diff
@@ -0,0 +1,27 @@
+    @NotBlank(message = "Поле event не должно быть пустым.")
+    Event event;
+    @NotNull(message = "Поле mark должно быть указано.")
+    @NotBlank(message = "Поле mark не должно быть пустым.")
```
Not blank применяется только к строкам, соотвественно в event и user тоже не нужно использовать (но их и не должно быть в таком виде в каком сейчас)

elbonreedZ commented on main-service/src/main/java/ru/practicum/rating/dto/RatingDto.java:
```diff
@@ -0,0 +1,20 @@
+@FieldDefaults(level = AccessLevel.PRIVATE)
+public class RatingDto {
+    Long id;
+    User user;
```
UserDto EventDto типы иначе ошибка lazy init

elbonreedZ commented on main-service/src/main/java/ru/practicum/rating/dto/UpdateRatingDto.java:
```diff
@@ -0,0 +1,26 @@
+@AllArgsConstructor
+@FieldDefaults(level = AccessLevel.PRIVATE)
+public class UpdateRatingDto {
+    @NotNull(message = "Поле id должно быть указано.")
```
в апдейт, как и в делит обычно указывают id в pathvariable

elbonreedZ commented on main-service/src/main/java/ru/practicum/rating/dto/UpdateRatingDto.java:
```diff
@@ -0,0 +1,26 @@
+    Long id;
+    @NotNull(message = "Поле user должно быть указано.")
+    @NotBlank(message = "Поле user не должно быть пустым.")
+    User user;
```
изменить пользователя оставившего лайк нельзя

elbonreedZ commented on main-service/src/main/java/ru/practicum/rating/dto/UpdateRatingDto.java:
```diff
@@ -0,0 +1,26 @@
+    @NotNull(message = "Поле user должно быть указано.")
+    @NotBlank(message = "Поле user не должно быть пустым.")
+    User user;
+    @NotNull(message = "Поле status должно быть указано.")
```
вообще при обновлении поля могут быть пустыми - если пустое, то его не обновляем. Но так как это единственное поле которые можно обновить, то оставить можно

elbonreedZ commented on main-service/src/main/java/ru/practicum/rating/mapper/RatingMapper.java:
```diff
@@ -0,0 +1,18 @@
+
+@Mapper(componentModel = MappingConstants.ComponentModel.SPRING)
+public interface RatingMapper {
+    Rating toRating(NewRatingDto newRatingDto);
```
тут сразу можно добавить Event event, User user в аргументы и соответсвующие аннотации

elbonreedZ commented on main-service/src/main/java/ru/practicum/rating/service/RatingService.java:
```diff
@@ -0,0 +1,19 @@
+public interface RatingService {
+    List<RatingDto> getAllMarksByUserId(long userId, RatingParam ratingParam);
+
+    RatingDto addRatingMark(long userId, @Valid NewRatingDto newRatingDto);
```
тут не нужна аннотация валид, в следующем тоже

no-vi-ka replied:
Здесь я написала вызов метода в контроллере и нажала создать его в сервисе, при этом скопировалась аннотация. Да, забыла её удалить.

elbonreedZ commented on main-service/src/main/java/ru/practicum/rating/service/RatingServiceImpl.java:
```diff
@@ -0,0 +1,84 @@
+        User user = userRepository.findById(userId).orElseThrow(() ->
+                new NotFoundException("User with id = " + userId + " not found."));
+        //Вместо этого добавила ограничение в таблице
+//        if (!ratingRepository.findAllByUser(user).stream()
```
оно кстати проще делается например метод existsByUser_IdAndEvent_Id

no-vi-ka replied:
Прописать ограничение в таблице мне кажется тоже достаточно просто.

elbonreedZ commented on main-service/src/main/java/ru/practicum/rating/service/RatingServiceImpl.java:
```diff
@@ -0,0 +1,84 @@
+//                .filter(rating -> rating.getEvent().getId().equals(newRatingDto.getEvent().getId())).collect(Collectors.toSet()).isEmpty()) {
+//            throw new ConditionsNotMetException(
+//                    "Rating mark for event with id = " + newRatingDto.getEvent().getId() + " from user with id = " + userId + " already exist.");
+//        }
```
и нужно провреить существование event

elbonreedZ replied:
и задать ивент

elbonreedZ commented on main-service/src/main/java/ru/practicum/rating/service/RatingServiceImpl.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    @Override
+    public void removeRatingMark(long userId, long ratingId) {
```
мне кажется тут не требуются такие проверки, лишний раз запросы отправлять. Если очень хочется проверить юсера, то можно сделать кастомный метод deleteByIdAndUser_Id @Modifying @Transactional

no-vi-ka committed:
Правки в соответствии с ревью команды.

no-vi-ka committed:
Добавить param в метод обновления.

no-vi-ka committed:
Добавить postman тесты для метода удаления оценок пользователей.

elbonreedZ committed:
Добавила тест на добавление рейтинга

elbonreedZ committed:
Поправила тест

Commit:
Работа с Rating теперь осуществляется по эндпоинту /users/{userId}/events/{eventId}/ratings

Так же в связи с этим был произведен большой рефакторинг кода

Реализована новая функциональность - поиск самых залайканных событий в сервисе (Публичная функция)

Написаны + переписаны POSTMAN тесты (все, кроме updateRating)

elbonreedZ committed:
Добавила тест на добавление рейтинга

elbonreedZ committed:
Добавила тесты на обновления, исправила чекстайл

elbonreedZ committed:
Поправила тест (в постмане проходит, на гитхабе нет)
